File: nodejs/src/gateways/fastpay.ts
/**
 * FastPay Gateway Client for Node.js
 * Handles Basic Auth and payment operations
 */

import axios, { AxiosInstance } from 'axios';
import { PaymentRequest, PaymentResponse, PaymentStatus, PaymentProvider } from '../index';

export interface FastPayConfig {
  storeId: string;
  password: string;
  baseUrl: string;
}

export interface FastPayPaymentRequest {
  amount: number;
  order_id: string;
  callback_url: string;
  webhook_url: string;
  metadata?: Record<string, string>;
}

export interface FastPayPaymentResponse {
  payment_id: string;
  status: 'pending' | 'completed' | 'failed' | 'expired' | 'cancelled';
  redirect_url: string;
  deep_link?: string;
}

export interface FastPayRefundRequest {
  amount: number;
  reason?: string;
}

export interface FastPayRefundResponse {
  refund_id: string;
  status: 'pending' | 'completed' | 'failed';
  amount: number;
}

export class FastPayClient {
  private client: AxiosInstance;
  private authHeader: string;

  constructor(config: FastPayConfig) {
    // Create Basic Auth header
    this.authHeader = `Basic ${Buffer.from(
      `${config.storeId}:${config.password}`
    ).toString('base64')}`;

    this.client = axios.create({
      baseURL: config.baseUrl,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': this.authHeader,
        'User-Agent': 'MesopotamiaSDK/0.1.0',
      },
      timeout: 30000,
    });
  }

  /**
   * Validate amount for FastPay (500 - 10M IQD)
   */
  private validateAmount(amount: number): void {
    const min = 500;
    const max = 10000000;
    if (amount < min || amount > max) {
      throw new Error(`Amount must be between ${min} and ${max} IQD`);
    }
  }

  /**
   * Create a payment transaction
   */
  async createPayment(request: PaymentRequest): Promise<PaymentResponse> {
    this.validateAmount(request.amount);

    const body: FastPayPaymentRequest = {
      amount: request.amount,
      order_id: request.orderId,
      callback_url: request.callbackUrl,
      webhook_url: request.webhookUrl,
      metadata: request.metadata,
    };

    try {
      const response = await this.client.post<FastPayPaymentResponse>(
        '/payments',
        body
      );

      const data = response.data;

      return {
        transactionId: data.payment_id,
        redirectUrl: data.redirect_url,
        deepLink: data.deep_link,
        status: this.mapStatus(data.status),
        provider: PaymentProvider.FASTPAY,
      };
    } catch (error: any) {
      throw new Error(`FastPay payment creation failed: ${error.message}`);
    }
  }

  /**
   * Get payment status
   */
  async getPaymentStatus(transactionId: string): Promise<PaymentStatus> {
    try {
      const response = await this.client.get<{status: string}>(
        `/payments/${transactionId}/status`
      );
      return this.mapStatus(response.data.status);
    } catch (error: any) {
      throw new Error(`FastPay status check failed: ${error.message}`);
    }
  }

  /**
   * Refund a payment
   */
  async refund(
    transactionId: string,
    amount: number,
    reason?: string
  ): Promise<FastPayRefundResponse> {
    try {
      const body: FastPayRefundRequest = { amount, reason };
      const response = await this.client.post<FastPayRefundResponse>(
        `/payments/${transactionId}/refund`,
        body
      );
      return response.data;
    } catch (error: any) {
      throw new Error(`FastPay refund failed: ${error.message}`);
    }
  }

  /**
   * Verify webhook signature (HMAC-SHA256)
   */
  verifyWebhook(signature: string, payload: string, secret: string): boolean {
    const crypto = require('crypto');
    const sigBytes = signature.startsWith('sha256=')
      ? signature.slice(7)
      : signature;

    const expected = crypto
      .createHmac('sha256', secret)
      .update(payload)
      .digest('hex');

    return crypto.timingSafeEqual(
      Buffer.from(sigBytes),
      Buffer.from(expected)
    );
  }

  private mapStatus(status: string): PaymentStatus {
    switch (status) {
      case 'completed':
        return PaymentStatus.COMPLETED;
      case 'failed':
        return PaymentStatus.FAILED;
      case 'expired':
        return PaymentStatus.EXPIRED;
      case 'cancelled':
        return PaymentStatus.CANCELLED;
      default:
        return PaymentStatus.PENDING;
    }
  }
}

export default FastPayClient;

File: core/src/fastpay.rs
// FastPay gateway implementation

use crate::client::HttpClient;
use crate::crypto::generate_hmac_sha256;
use crate::error::{MesopotamiaError, Result};
use crate::types::{Environment, PaymentProvider, PaymentRequest, PaymentResponse, PaymentStatus};
use reqwest::header::HeaderMap;
use serde::Serialize;
use std::collections::HashMap;

/// FastPay gateway client
pub struct FastPayClient {
    base_url: String,
    store_id: String,
    password: String,
    client: HttpClient,
}

impl FastPayClient {
    /// Create a new FastPay client
    pub fn new(base_url: String, store_id: String, password: String, timeout_ms: u64) -> Self {
        Self {
            base_url,
            store_id,
            password,
            client: HttpClient::new(timeout_ms),
        }
    }

    /// Create a payment transaction
    pub async fn create_payment(&self, request: &PaymentRequest) -> Result<PaymentResponse> {
        // Validate amount range
        let (min, max) = PaymentProvider::FastPay.amount_range();
        if request.amount < min || request.amount > max {
            return Err(MesopotamiaError::InvalidAmount {
                min,
                max,
                provided: request.amount,
            });
        }

        // Validate order ID
        if request.order_id.len() > 50 {
            return Err(MesopotamiaError::InvalidOrderId {
                reason: "Order ID must be max 50 characters".to_string(),
            });
        }

        // Build request body
        let body = FastPayInitRequest {
            amount: request.amount,
            order_id: request.order_id.clone(),
            callback_url: request.callback_url.clone(),
            webhook_url: request.webhook_url.clone(),
            description: request.description.clone().unwrap_or_default(),
            metadata: request.metadata.clone().unwrap_or_default(),
        };

        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .post(format!("{}/payment/init", self.base_url))
                .basic_auth(&self.store_id, Some(&self.password))
                .json(&body)
                .send()
                .await?;

            Ok(resp)
        }).await?;

        // Handle response
        let status = response.status();
        let response_text = response.text().await?;

        if status.is_success() {
            let init_response: FastPayInitResponse = serde_json::from_str(&response_text)
                .map_err(|e| MesopotamiaError::SerializationError(e))?;

            Ok(PaymentResponse {
                transaction_id: init_response.payment_id,
                redirect_url: init_response.redirect_url,
                deep_link: None,
                status: PaymentStatus::Pending,
                provider: PaymentProvider::FastPay,
            })
        } else {
            self.handle_error(status, &response_text)
        }
    }

    /// Get payment status
    pub async fn get_payment_status(&self, payment_id: &str) -> Result<PaymentStatus> {
        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .get(format!("{}/payment/status/{}", self.base_url, payment_id))
                .basic_auth(&self.store_id, Some(&self.password))
                .send()
                .await?;

            Ok(resp)
        }).await?;

        let status_response: FastPayStatusResponse = response.json().await?;
        Ok(parse_status(&status_response.status))
    }

    /// Process refund
    pub async fn refund(&self, payment_id: &str, amount: u64, reason: Option<String>) -> Result<FastPayRefundResponse> {
        let body = FastPayRefundRequest {
            payment_id: payment_id.to_string(),
            amount,
            reason: reason.unwrap_or_default(),
        };

        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .post(format!("{}/payment/refund", self.base_url))
                .basic_auth(&self.store_id, Some(&self.password))
                .json(&body)
                .send()
                .await?;

            Ok(resp)
        }).await?;

        let refund_response: FastPayRefundResponse = response.json().await?;
        Ok(refund_response)
    }

    /// Verify webhook signature
    pub fn verify_webhook(&self, signature: &str, payload: &str) -> bool {
        generate_hmac_sha256(&self.password, payload)
            .map(|sig| format!("sha256={}", sig) == signature)
            .unwrap_or(false)
    }

    /// Handle error response from gateway
    fn handle_error(&self, status: reqwest::StatusCode, body: &str) -> Result<PaymentResponse> {
        if let Ok(error_resp) = serde_json::from_str::<FastPayErrorResponse>(body) {
            Err(MesopotamiaError::from_gateway_code(
                "fastpay",
                &error_resp.code,
                &error_resp.message,
            ))
        } else {
            Err(MesopotamiaError::GatewayError {
                provider: "fastpay".to_string(),
                code: status.as_u16().to_string(),
                message: body.to_string(),
            })
        }
    }
}

/// FastPay request/response types
#[derive(Debug, Serialize)]
struct FastPayInitRequest {
    amount: u64,
    order_id: String,
    callback_url: String,
    webhook_url: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    description: String,
    metadata: HashMap<String, String>,
}

#[derive(Debug, Deserialize)]
struct FastPayInitResponse {
    payment_id: String,
    redirect_url: String,
    status: String,
    expires_at: String,
}

#[derive(Debug, Deserialize)]
struct FastPayStatusResponse {
    payment_id: String,
    status: String,
    amount: u64,
    order_id: String,
}

#[derive(Debug, Serialize)]
struct FastPayRefundRequest {
    payment_id: String,
    amount: u64,
    reason: String,
}

#[derive(Debug, Deserialize)]
pub struct FastPayRefundResponse {
    pub refund_id: String,
    pub payment_id: String,
    pub amount: u64,
    pub status: String,
    pub created_at: String,
}

#[derive(Debug, Deserialize)]
struct FastPayErrorResponse {
    error: String,
    code: String,
    message: String,
}

/// Parse payment status from string
fn parse_status(status: &str) -> PaymentStatus {
    match status.to_lowercase().as_str() {
        "pending" => PaymentStatus::Pending,
        "completed" => PaymentStatus::Completed,
        "failed" => PaymentStatus::Failed,
        "cancelled" => PaymentStatus::Cancelled,
        "expired" => PaymentStatus::Expired,
        "refunded" => PaymentStatus::Cancelled, // Treat refunded as cancelled for now
        _ => PaymentStatus::Pending,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    #[test]
    fn test_parse_status() {
        assert_eq!(parse_status("pending"), PaymentStatus::Pending);
        assert_eq!(parse_status("completed"), PaymentStatus::Completed);
        assert_eq!(parse_status("FAILED"), PaymentStatus::Failed);
        assert_eq!(parse_status("refunded"), PaymentStatus::Cancelled);
    }

    #[test]
    fn test_client_creation() {
        let client = FastPayClient::new(
            "https://test.com".to_string(),
            "store_123".to_string(),
            "pass".to_string(),
            30000,
        );
        assert_eq!(client.store_id, "store_123");
        assert_eq!(client.base_url, "https://test.com");
    }
}
